From ce6155625318548fd5a4ed2b232dfb7abaa52797 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 11:59:34 -0400
Subject: [ethernet] mlx4: Implement the SIOCGHWTSTAMP ioctl

Message-id: <465cd113ec34a1c3ae6826f31e7d1181a880c31b.1410425016.git.avadai@redhat.com>
Patchwork-id: 93592
O-Subject: [Patch RHEL7.1 V1 038/197] mlx4_en: Implement the SIOCGHWTSTAMP ioctl
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Ben Hutchings <bhutchings@solarflare.com>

Compile-tested only.

Signed-off-by: Ben Hutchings <bhutchings@solarflare.com>
(cherry picked from commit 100dbda8e40bf4b537332d909660ddf9945196ff)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_netdev.c b/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
index f038964..b68dde0 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
@@ -2105,7 +2105,7 @@ static int mlx4_en_change_mtu(struct net_device *dev, int new_mtu)
  return 0;
 }
 
-static int mlx4_en_hwtstamp_ioctl(struct net_device *dev, struct ifreq *ifr)
+static int mlx4_en_hwtstamp_set(struct net_device *dev, struct ifreq *ifr)
 {
  struct mlx4_en_priv *priv = netdev_priv(dev);
  struct mlx4_en_dev *mdev = priv->mdev;
@@ -2164,11 +2164,21 @@ static int mlx4_en_hwtstamp_ioctl(struct net_device *dev, struct ifreq *ifr)
        sizeof(config)) ? -EFAULT : 0;
 }
 
+static int mlx4_en_hwtstamp_get(struct net_device *dev, struct ifreq *ifr)
+{
+ struct mlx4_en_priv *priv = netdev_priv(dev);
+
+ return copy_to_user(ifr->ifr_data, &priv->hwtstamp_config,
+       sizeof(priv->hwtstamp_config)) ? -EFAULT : 0;
+}
+
 static int mlx4_en_ioctl(struct net_device *dev, struct ifreq *ifr, int cmd)
 {
  switch (cmd) {
  case SIOCSHWTSTAMP:
-  return mlx4_en_hwtstamp_ioctl(dev, ifr);
+  return mlx4_en_hwtstamp_set(dev, ifr);
+ case SIOCGHWTSTAMP:
+  return mlx4_en_hwtstamp_get(dev, ifr);
  default:
   return -EOPNOTSUPP;
  }
-- 
1.7.1